### 8.3 --- Parcours d'un tableau

En utilisant le tableau `pda` de la partie précédente, proposer un programme permettant d'obtenir le nombre de français ayant entre 1 et 8 ans.

In [1]:
pda = [691754, 714585, 723589, 741105, 761638, 782397, 806296, 813727, 830361, 836626]
pda[1] + pda[2] + pda[3] + pda[4] + pda[5] + pda[6] + pda[7] + pda[8]

6173698

Comment s'y prendre quand le tableau est beaucoup plus long ?

Une meilleure solution consiste à utiliser une **boucle** pour parcourir les cases du tableau concernées en **accumulant** le nombre total dans une variable.

On commence par initialiser la variable `n` à `0` :

```python
n = 0
```

Puis on effectue une boucle `for` donnant successivement à la variable `i` toutes les valeurs entre 1 et 8, et on ajoute à chaque fois la valeur `pda[i]` à la variable `n` :

```python
for i in range(1, 9):
    n += pda[i]
```

On peut vérifier qu'après la boucle la variable `n` contient bien la valeur précédente.


In [3]:
n = 0
for i in range (1, 9):
    n += pda[i]

n

6173698

Écrire un programme qui demande à l'utilisateur

- un âge minimal
- un âge maximal

et qui renvoie le nombre de français dont l'âge est compris entre le minimal (inclu) et le maximal (exclu).

Par exemple :

        >>> âge minimum (inclu) : 1
        >>> âge maximum (exclu) : 9
        il y a 6173698 personnes qui ont entre 1 et 9 ans

Pour parcourir **toutes les cases** d'un tableau `t`, on peut utiliser une  boucle `for` allant de 0 inclus à `len(t)` exclu.

Proposer un programme qui affiche le nombre total de français contenus dans votre tableau `pda`.

In [7]:
population_totale = 0
for i in range (0, len(pda)):
    population_totale += pda[i]

print(population_totale)

7702078


### 8.4 --- Construire de grands tableaux

Pour construire des tableaux d'une taille arbitraire, on peut utiliser la syntaxe :

```python
>>> t = [0] * 1000
```

On donne la taille du tableau après le symbole `*` (ici `1000`) et entre crochet une valeur qui sera donnée à toutes les cases du tableau (ici `0`).

On obtient donc un tableau de taille 1000 dont toutes les cases contiennent pour l'instant la valeur `0`.

Créer un tel tableau et vérifier sa taille.

In [8]:
t = [0] * 1000
len(t)

1000

Une fois le tableau créé, on peut le **remplir** avec des valeurs de son choix.

Remplir le tableau `t` avec la valeur des carrés des 1000 premiers entiers.

In [11]:
for i in range(0, 1000):
    t[i] = i * i

100

**Attention**, l'opération `[0]*1000` n'implique par une multiplication (même si on voit le symbole `*`). Il s'agit là d'une opération spécifique aux tableaux.


Il est possible de *concaténer* deux tableaux, c'est-à-dire de construire un nouveau tableau contenant bout à bout les éléments des deux tableaux. On le fait avec l'opération `+` :

```python
>>> [8, 13, 21] + [34, 55]
[8, 13, 21, 34, 55]
```

#### Tableaux et chaînes de caractères

Les chaînes de caractères offrent certaines ressemblance avec les tableaux.

- taille d'une chaîne de caractère : `len(ch)`
- accès au i-ème caractère d'une chaîne : `ch[i]`
- concaténation : `+`
- répétition d'un valeur : `*`

1. Créer la variable `ch` initialisée à la chaîne de caractères `"bonjour"`.
2. Afficher la longueur de la chaîne. 
3. Vérifier que le 3ème caractère de la chaîne est bien le `n`
4. Concaténer cette chaîne avec la deuxième chaîne `" le monde !"`

In [15]:
ch = "bonjour"
print(len(ch))
print(ch[2])
ch = ch + " le monde !"
print(ch)

7
n
bonjour le monde !


**Attention**, contrairement aux tableaux, les caractères d'une chaîne ne peuvent pas être modifiés. On dit qu'une chaîne est immuable.

In [18]:
ch[2] = "Z"

TypeError: 'str' object does not support item assignment